pkg/services/provisioning/datasources/datasources.go

```diff
@@ -153,6 +154,28 @@ func (dc *DatasourceProvisioner) applyChanges(ctx context.Context, configPath st
 		}
 	}
 
+	prunableProvisionedDataSources, err := dc.store.GetPrunableProvisionedDataSources(ctx, &datasources.GetPrunableProvisionedDataSourcesQuery{})
+	if err != nil {
+		return err
+	}
+
+	staleProvisionedDataSources := []*deleteDatasourceConfig{}
+
+	for _, prunableProvisionedDataSource := range prunableProvisionedDataSources {
+		key := DataSourceMapKey{
+			OrgId: prunableProvisionedDataSource.OrgID,
+			Name:  prunableProvisionedDataSource.Name,
+		}
+		if _, ok := willExistAfterProvisioning[key]; !ok {
+			staleProvisionedDataSources = append(staleProvisionedDataSources, &deleteDatasourceConfig{OrgID: prunableProvisionedDataSource.OrgID, Name: prunableProvisionedDataSource.Name})
+			willExistAfterProvisioning[key] = false
```

Comment:
You are not using the provision path in order to differ if a datasource should be deleted or not. I don't think it's actually needed. You can just use `is_prunable` and remove `provisionedFrom` from the model. A datasource marked to be pruned can be deleted whenever it does not appear anymore, regardless of the provision file path.